src/parser/ts/index.ts

```diff
@@ -0,0 +1,528 @@
+  return items.reduce((res, filter) => {
+    let value: any;
+    if (isByIdsQuery(filter.value)) {
+      value = filter.value.map(item => flattenServerQueryFilters([item]));
+    } else if(isSTSFilterArr(filter.value)) {
+      value = {...res[filter.path], ...flattenServerQueryFilters(filter.value)};
+    } else if (isArray(filter.value)){
+      value = filter.value.map(flattenServerQueryFilters);
+    } else {
+      value = filter.value
+    }
+    res[filter.path] = value;
+    return res;
+  }, {});
+}
+
+export const walk = (
+  route: STSItem,
+  iterator: (item: STSQueryItem | STSFunction) => any
+) => {
+  if (isArray(route)) {
+    return route.map(r => walk(r, iterator));
+  }
+  const resArr = [];
+  const res = iterator(route);
+  if (res !== undefined) {
+    resArr.push(res);
+  }
+  if (isSTSFunction(route)) {
+    let args: STSItem;
+    if(isMapFunction(route)) {
+      args = route.args[0];
+    } else {
+      args = route.args;
+    }
+    resArr.push(...walk(args, iterator));
+  }
+  return resArr;
+};
+
+export const getIndicesOfMultiaryFunctionArgs = (route: STSQueryItem[] | STSFunction): number[] => {
+  const responseArr = [];
+  const argsIndices = [];
+  let index = 0;
+  walk(route, obj => {
+    if (isMultiaryFunction(obj) && obj.args.length === 1) {
+      argsIndices.push(...obj.args);
+    } else if (isSTSReference(obj) && !hasIdsFilter(obj)) {
+      if (argsIndices.includes(obj)) {
+        responseArr.push(index);
+      }
+      index++;
+    }
+  });
+  return responseArr;
+};
+
+export const getServerFilters = (
+  route: STSQueryItem[] | STSFunction
+): StringMap[] => {
+  const responseArr: StringMap[] = [];
+  const filter = (_, __, parent) => isServerFilter(parent);
+  walk(route, obj => {
+    if (isSTSReference(obj) && !hasIdsFilter(obj)) {
+      const serverQuery = filterDeep(filterDeep(obj.query, filter), filter);
+      const flatFilters = flattenServerQueryFilters(serverQuery || []);
+      responseArr.push(flatFilters);
+    }
+  });
+  return responseArr;
+};
+
+export function flattenClientQueryFilters(filters: STSFilter[], path = []): STSClientFilter[] {
+  const res: STSClientFilter[] = [];
+  for(const filter of filters) {
+    if (isServerFilter(filter) && isSTSFilterArr(filter.value)) {
+      res.push(
+        ...flattenClientQueryFilters(filter.value, [...path, filter.path])
+      )
+    } else {
+      res.push({
+        path: [...path, filter.path].join('.'),
+        value: filter.value,
+        filter: filter.filter
+      } as STSClientFilter)
+    }
+  }
+  return res;
+}
+
+export const getClientFilters = (route: STSQueryItem[] | STSFunction): STSClientFilter[][] => {
+  const responseArr: STSClientFilter[][] = [];
+  const filter = (_, __, parent) => isClientFilter(parent) || isArray(parent.value);
+  walk(route, obj => {
+    if (isSTSReference(obj) && !hasIdsFilter(obj)) {
+      const clientQuery = filterDeep(filterDeep(obj.query, filter) || [], filter);
```

Comment:
to avoid doing double deep filtering like:
```typescript
filterDeep(filterDeep(obj.query, filter) || [], filter);
```

we can replace : 
```typescript
const filter = (_, __, parent) => isClientFilter(parent) || isArray(parent.value);
```
with:
```typescript
const filter = (value, _, parent) => isClientFilter(parent) || (isArray(value) && value.some(isClientFilter));
```

Reply:
I wish it worked, but it breaks tests actually :(

Reply:
decided to keep as it is